---
title: Repeated loss of variation in insect ovary morphology highlights the role of developmental constraint in life-history evolution - Supplementary Methods
output: 
  bookdown::pdf_document2:
    keep_tex: yes
    latex_engine: pdflatex
    toc: no
  #bookdown::html_document2: default
bibliography: manuscript_supplement.bib
csl: nature.csl
link-citations: yes
header-includes:
 \usepackage{float}
 \floatplacement{figure}{H}
---

```{r global options, include = FALSE}
knitr::opts_chunk$set(echo=FALSE, include = FALSE, warning=FALSE, message=FALSE, cache=TRUE)
```

```{r preliminaries,cache=FALSE}
library(BAMMtools)
library(phytools)
library(coda)
library(OUwie)
library(corHMM)
library(grid)
library(gridExtra)
library(gtable)
library(geiger)
library(arbutus)
library(dplyr)
library(ggtree)
library(ape)
library(ggplot2)
library(RColorBrewer)
theme_set(theme_classic())
options(tinytex.clean = FALSE)
```

\makeatletter
\renewcommand{\thefigure}{S\@arabic\c@figure}
\renewcommand{\thetable}{S\@arabic\c@table}
\makeatother

```{r load_data, cache=FALSE}
# This script reads in and formats the ovariole number dataset
source("analysis/get_ovariole_number_data.R")
```

Samuel H. Church^1^\*, Bruno A. S. De Medeiros^1,2^, Seth Donoughe^1,3^, Nicole L. MÃ¡rquez Reyes^4^, Cassandra G. Extavour^1,5^\*


^1^ Department of Organismic and Evolutionary Biology, Harvard University, Cambridge, MA 02138, USA

^2^ Smithsonian Tropical Research Institute, Panama City, Panama

^3^ Department of Molecular Genetics and Cell Biology, University of Chicago, Chicago, IL 60637, USA

^4^ Department of Biology, Universidad de Puerto Rico en Cayey, Cayey 00736, PR

^5^ Department of Molecular & Cellular Biology, Harvard University, Cambridge, MA 02138, USA

\* Corresponding author

```{=latex}
% Trigger ToC creation in LaTeX
\tableofcontents
```

# Gathering ovariole number records {#sec:gathering}

We searched the published literature for references to insect ovariole number using a predetermined set of 131 search terms, entered into Google Scholar (scholar.google.com) between June and October of 2019. Each search term consisted of an insect taxonomic group and the words "ovariole number". This list was created to include all insect orders, many large insect families, and groups well-represented in the insect egg dataset. The list of search terms is available in the supplementary file 'ovariole_number_search_terms.tsv'.

For each search term, we evaluated all publications in the first page of results (ten publications). For 61 search terms that had a large number of informative hits, significant representation in the egg dataset, or that corresponded to very speciose groups, we evaluated an additional 20 publications. If a publication reported ovariole number for one or more insect species, we recorded the following information: (1) genus, (2) species name, when available, (3) taxonomic order, (4) sample size, when available, (5) ovariole number, and (6) additional notes (e.g. for eusocial insects, whether the observation was made in a reproductive or non-reproductive individual). This dataset is made publically available at Dryad (doi:10.5061/dryad.59zw3r253).

Ovariole number was recorded as either an average with deviations, a range, or a single total value, with priority for recording given in that order when multiple data types were available. Ovariole number was recorded as the total number of ovarioles per female, summing over both the left and right adult ovaries. When authors reported ovariole number from a single ovary, the total value was calculated by doubling the reported value. When authors described differences between the two ovaries, this information was recorded in an additional column.

Using this approach, we gathered `r nrow(raw_ON)` records for ovariole number from 460 publications. A full list of publications is provided in the supplementary files 'ovariole_number_bibliography.pdf'. We matched the scientific names to additional taxonomic information using the software TaxReformer[@TaxReformer] and found additional taxonomic data for `r nrow(matched)` of the `r nrow(raw_ON)` records. We verified that TaxReformer had found a valid match by comparing the originally recorded taxonomic order to the order populated by online databases, and removed `r nrow(raw_ON %>% filter(problem == "taxonomy"))` taxonomic records for which these values did not match. For all subsequent analyses, we also excluded observations made in non-reproductive individuals from eusocial species (workers), as well as two observations which represented significant outliers and could not be validated using additional sources or figures[@su2015testicular;@hernandez2017biology].


# Phylogenetic trees {#sec:trees}

The analyses herein were performed using the insect phylogeny published in Church et al, 2019[@church2019insect], unless otherwise specified. This phylogeny was constructed by combining ribosomal genetic data from `r length(tree$tip.label)` insect genera, published originally in the SILVA database[@Yilmaz2014], with constrained, time-calibrated nodes for each insect order, published originally in Misof et al, 2014[@Misof:2014jv]. This phylogeny is enriched for insect genera with records in the egg trait dataset, and also has considerable overlap with the genera included in this ovariole number dataset (`r ON %>% filter(genus %in% tree$tip.label) %>% distinct(genus) %>% nrow()` genera). For generalized least squares analyses and trait model comparisons, analyses were performed over a posterior distribution of trees associated with this published phylogeny[@church2019insect].

Analyses of insect family-level ovariole number, egg size, and body size were performed using the insect phylogeny published in Rainford et. al, 2014[@rainford2014phylogenetic].

Analyses of Drosophilidae ovariole number, egg size, and body size were performed using a phylogeny newly assembled for this study. Published genetic data for 317 Drosophilidae species were retrieved from NCBI in June of 2019[@katoh2017multiple;@magnacca2015rapid;@lapoint2014phylogenetics;@o2011phylogenetic;@lapoint2011phylogenetic;@bonacum2005phylogeny;@o2004phylogenetic;@lapoint2013diversification;@baker1997multiple]. These data encompassed 41 gene regions including mitochondrial, nuclear, and ribosomal genes. When multiple sequences for a gene region were available from the same species, the one with the least amount of missing data was selected. Each gene region was aligned using the program MAFFT[@Katoh2013], model auto selected). Alignments were concatenated and trimmed to 3% occupancy across species using the program phyutlity[@smith2008phyutility]. Documentation including accession numbers, sequence files, and alignments are available in the supplementary directory 'https://github.com/shchurch/insect_ovariole_number_evolution_2020/phylogeny/Drosophilidae_sequences/'.

To the extent possible, sequence data were not curated beyond what was downloaded from NCBI, with the following exceptions: [1] two sequences labeled as 16S that did not align to other 16S sequences were removed manually. [2] COI sequences were trimmed to remove regions with large quantities of missing sites prior to alignment. [3] One species name (_D. albovittata_) was corrected for typographical error. [4] Sequences identified as _Drosophila crassifemur_ were taxonomically corrected to _Scaptomyza crassifemur_[@o2003placement].

Phylogenetic estimation of the Drosophilidae data were performed using RAxML (model GTRGAMMA), setting the split between Hawaiian _Drosophila_ and _Scaptomyza_ as the root of the tree[@katoh2017multiple;@o2011phylogenetic]. The final tree was pruned to remove undescribed species (e.g. _Drosophila_ nr _dorsigera_), and was time-calibrated using the R package ape, function chronos (default parameters, version `r packageVersion('ape')`)[@paradis2004ape]. This tree is available in the supplementary file 'https://github.com/shchurch/insect_ovariole_number_evolution_2020/phylogeny/Drosophilidae_time_calibrated.tre'.

# Phylogenetic regressions

## Combining datasets {#sec:combining}

We combined the data we collected on total ovariole number with existing datasets of egg size and shape[@church2019dataset], insect lifetime fecundity and adult dry body weight[@Gilbert2007], average adult body length per insect family[@rainford2016phylogenetic], and several lineage-specific measures of adult body size[@iwata1966large;@Iwata1966a;@waloff1954number;@starmer2003phylogenetic;@reinhardt2005low].

Ovariole number and egg size[@church2019insect] data were combined by matching records at the species level (Fig 2a). When multiple records existed for a given species, the dataset was randomly shuffled and a single matching record was selected. This variation across records for the same species was accounted for in regressions by reshuffling and matching records at each iteration of the analysis. In order to test whether results were robust with a larger sample size, we also matched records at the genus level following the same reshuffling method (Fig. \ref{fig:PGLS-egg-ON-genus}).

Average adult body length per insect family[@rainford2016phylogenetic] was matched to the average ovariole number for the corresponding families (Fig. \ref{fig:PGLS-egg-ON-family}). The Rainford et al, 2016[@rainford2016phylogenetic] dataset contains a small number of average adult body lengths at the order level (e.g. Strepsiptera), which were matched to their equivalent group in the ovariole number dataset. To test the effect of uncertainty in the estimated average ovariole number on our results, the dataset for each family was downsampled by half at each iteration of the regression analysis.

Ovariole number, egg volume, lifetime fecundity, and adult body weight[@Gilbert2007] were combined by matching records at the species level and genus level, using the same method as described above (Figs. 2b, 3, and \ref{fig:PGLS-Gilbert-name}). We excluded one value from this dataset which appeared to include a typographical error for lifetime fecundity (Hymenoptera: Trichogrammatidae, lifetime fecundity recorded as 0.1[@Gilbert2007]).

Several lineage-specific measurements for body size were matched to the ovariole number and egg size datasets, as follows: Drosophilidae thorax length[@starmer2003phylogenetic] was matched at the species level (Fig. 2c), Orthoptera body length[@waloff1954number] was matched at the genus level (Fig. 2d), Hymenoptera mesosoma width[@Iwata1966a] was matched at the genus level, and Curculonoidea elytra length[@iwata1966large] was matched at the genus level (Fig. \ref{fig:PGLS-Coleoptera-Hymenoptera}).

## Phylogenetic Generalized Least Squares (PGLS) analyses {#sec:PGLS}

```{r PGLS-posterior, include=TRUE, cache=FALSE}
#source("analysis/run_pgls_over_posterior_distribution.R")
load("analysis/pgls_results_June2020/run_pgls_over_posterior_distribution.RData")

# analysis names in table 
pgls_analysis_names <- c("ovariole number vs egg volume",
  "ovariole number vs egg volume",
  "ovariole number vs egg volume, residuals to body length",
  "ovariole number vs egg volume, residuals to body length, excluding Trigonalidae",
  "ovariole number vs egg volume, residuals to body weight",
  "ovariole number vs egg volume, residuals to body weight",
  "ovariole number vs lifetime fecundity",
  "ovariole number vs lifetime fecundity",
  "Curculionoidea ovariole number vs egg volume, residual to elytra length",
  "Orthoptera ovariole number vs egg volume, residuals to body length",
  "Hymenoptera ovariole number vs egg volume, residuals to mesosoma width",
  "Drosophilidae ovariole number vs egg volume, residuals to thorax length")
# level at which each PGLS was performed 
taxonomic_levels <- c("species","genus","family","family","species","genus","species","genus","genus","genus","genus","species")

sp <- " -- " # separator in tables

# write table
all_table <- as.data.frame(do.call(rbind, allometry_tables)) %>% round(.,3) %>% mutate(analysis = pgls_analysis_names, "taxonomic level" = taxonomic_levels,slope = paste(sprintf("%.3f",slope_min),sprintf("%.3f",slope_max),sep=sp),
      intercept = paste(sprintf("%.3f",int_min),sprintf("%.3f",int_max),sep=sp),
      "p-value" = paste(sprintf("%.3f",pval_min),gsub("0.000","<0.000",sprintf("%.3f",pval_max)),sep=sp),
      "taxa" = taxa) 
knitr::kable(all_table  %>% select(c("analysis","taxonomic level","slope","intercept","p-value","taxa")), "latex",caption="Results of PGLS analysis of ovariole number, egg size, and fecundity across a posterior distribution", booktabs = T, linesep = "") %>%
  kableExtra::kable_styling(latex_options = "striped") %>%
  kableExtra::column_spec(1, width = "16em") %>%
  kableExtra::kable_styling(latex_options = "hold_position")  %>%
  kableExtra::kable_styling(font_size = 8) 
```  

```{r PGLS-body-posterior, include=TRUE, cache=FALSE}
#source("analysis/run_pgls_body_ovariole_number.R")
load("analysis/pgls_results_June2020/run_pgls_body_ovariole_number.RData")

# analysis names in table 
pgls_analysis_names <- c("ovariole number vs body length",
  "ovariole number vs body weight",
  "ovariole number vs body weight",
  "Curculionoidea ovariole number vs elytra length",
  "Orthoptera ovariole number vs body length",
  "Hymenoptera ovariole number vs mesosoma width",
  "Drosophilidae ovariole number vs thorax length")
# level at which each PGLS was performed 
taxonomic_levels <- c("family","species","genus","genus","genus","genus","species")

# write table 
body_all_table <- as.data.frame(do.call(rbind, body_allometry_tables)) %>% round(.,3) %>% mutate(analysis = pgls_analysis_names, "taxonomic level" = taxonomic_levels,slope = paste(sprintf("%.3f",slope_min),sprintf("%.3f",slope_max),sep=sp),
       intercept = paste(sprintf("%.3f",int_min),sprintf("%.3f",int_max),sep=sp),
       "p-value" = paste(sprintf("%.3f",pval_min),gsub("0.000","<0.000",sprintf("%.3f",pval_max)),sep=sp),
       "taxa" = taxa)
knitr::kable(body_all_table %>% select(c("analysis","taxonomic level","slope","intercept","p-value","taxa")), "latex",caption="Results of PGLS analysis of ovariole number and body size across a posterior distribution", booktabs = T, linesep = "") %>%
  kableExtra::kable_styling(latex_options = "striped") %>%
  kableExtra::column_spec(1, width = "16em") %>%
  kableExtra::kable_styling(latex_options = "hold_position")  %>%
  kableExtra::kable_styling(font_size = 8) 
```

For each regression, we repeated a Phylogenetic Generalized Least Squares (PGLS) analysis 1000 times, accounting for phylogenetic and phenotypic uncertainty, using the R packages ape (version `r packageVersion('ape')`)[@paradis2004ape] and nlme (version `r packageVersion('nlme')`)[@pinheiro2014r]. In these analyses we used a Brownian Motion based covariance matrix for traits. 

For regressions at the species and genus level, we reshuffled and matched records at each iteration to account for variation across records for the same taxon. For regressions at the family level we recalculated the average ovariole number per insect family, downsampling the representation for each family by half. No posterior distribution was made available with the published family level phylogeny[@rainford2014phylogenetic].

To account for body size, we calculated the phylogenetic residuals[@revell2010phylogenetic] of each trait to body size, and then compared the evolution of these residuals using a PGLS regression.

```{r PGLS-setup}
# build body size dataset
# set factor for resampling the body size dataframe, here no downsampling used
downsample_factor <- 1.0
fam_count_threshold <- 1
source("analysis/combine_body_egg_datasets.R")
source("analysis/pgls_functions.R") # get functions for plotting PGLS results
```

```{r PGLS-egg-ON-name}
### PGLS egg volume vs ON, species level pgls
egg_ON_name_pgls <- egg_ON_name %>% select(genus,vol,on,group) %>% rename(rank = genus, trait1 = on, trait2 = vol)

# this function helps format scientific notation correctly
scientific_10 <- function(x) {
  parse(text=gsub("1e", "10^", scales::scientific_format()(x)))
}

pdf("manuscript_figures/panels/FigEgg_ON_name.pdf",height=5,width=7.5,useDingbats = F)
  g1 <- ggplot(egg_ON_name_pgls,aes(x = trait1, y = trait2, color = group)) +
      geom_point() +
      scale_color_manual(values = mrk) +
      scale_x_log10(breaks=c(1,10,100,1000,10000)) +
      scale_y_log10(label=scientific_10,breaks=c(0.00001,0.0001,0.001,0.01,0.1,1,10,100,100)) +
      theme(legend.position = "none") +
      xlab('ovariole number') +
      ylab(bquote('egg volume '(~mm^3)))
  print(g1)
invisible(dev.off())

### Gilbert egg volume vs ON, independent body weight, genus level pgls
egg_ON_gilbert_genus_pgls <- egg_ON_gilbert_genus %>% select(genus,logvol,logon,logweight,group) %>% mutate(rank = genus, trait1 = logon, trait2= logvol, indep = logweight)
pdf("manuscript_figures/panels/FigEgg_gilbert_weight_residuals.pdf",height=2.5,width=2.5,useDingbats = F)
run_plotting_resid_pgls(egg_ON_gilbert_genus_pgls,tree,"gilbert_genus") + xlab("residual of ovariole number") +
  ylab("residual of egg volume")
invisible(dev.off())

### Drosophilidae pgls
DIP_pgls <- DIP_body_egg_ON_name %>% select(name,logvol,logon,logbody,group) %>% rename(rank = name, trait1 = logon, trait2= logvol, indep = logbody)
pdf("manuscript_figures/panels/FigEgg_Drosophila_residuals.pdf",height=2.5,width=2.5,useDingbats = F)
run_plotting_resid_pgls(DIP_pgls,dros_tree,"DIP") + xlab("residual of ovariole number") +
  ylab("residual of egg volume")
invisible(dev.off())

### Orthoptera pgls
ORT_pgls <- ORT_body_egg_ON_genus %>% select(genus,logvol,logon,logbody,group) %>% rename(rank = genus, trait1 = logon, trait2= logvol, indep = logbody)
pdf("manuscript_figures/panels/FigEgg_Orthoptera_residuals.pdf",height=2.5,width=2.5,useDingbats = F)
run_plotting_resid_pgls(ORT_pgls,tree,"ORT") + xlab("residual of ovariole number") +
  ylab("residual of egg volume")
invisible(dev.off())
```

\clearpage

```{r PGLS-egg-ON-genus, fig.cap="**Egg volume vs ovariole number, matching records at the genus level.**", fig.width=7.5, fig.height=5, include=TRUE}
### PGLS egg volume vs ON, species level pgls
egg_ON_genus_pgls <- egg_ON_genus %>% select(genus,logvol,logon,group) %>% rename(rank = genus, trait1 = logon, trait2 = logvol)
run_plotting_pgls(egg_ON_genus_pgls,tree,"egg_ON_genus") +
  xlab(bquote('ovariole number,'~log[10])) +
  ylab(bquote('egg volume,'~mm^3~','~log[10]))
```

```{r PGLS-Gilbert-name, fig.cap="**Egg volume vs ovariole number, phylogenetic residuals to dry adult body weight, matching records at the species level.**", fig.width=3, fig.height=3, include=TRUE}
### Gilbert body weight, species level pgls
egg_ON_gilbert_name_pgls <- egg_ON_gilbert_name %>% select(genus,logvol,logon,logweight,group) %>% mutate(rank = genus, trait1 = logon, trait2= logvol, indep = logweight)
run_plotting_resid_pgls(egg_ON_gilbert_name_pgls,tree,"gilbert_name") + xlab("residual of ovariole number") +
  ylab("residual of egg volume")
```

```{r PGLS-egg-ON-family, fig.cap="**Family average egg volume vs ovariole number, phylogenetic residuals to adult body length.** Point outlined in black = Hymenoptera: Trigonalidae, an outlier that drives the relationship.",  fig.width=3, fig.height=3, include=TRUE}
### PGLS egg volume vs ON, family averages
body_egg_ON_family_pgls <- body_egg_ON_full %>% select(rank,logvol,logon,logbody,group) %>% rename(trait1 = logon, trait2= logvol, indep = logbody)
egfam1 <- run_plotting_resid_pgls(body_egg_ON_family_pgls,fam_tree,"body_egg_ON_family")

egfam1 + xlab("residual of average ovariole number") +
  ylab("residual of average egg volume") +
  geom_point(data = (egfam1$data %>% filter(rank == "Trigonalidae")),aes(fill=group),color="black",stroke=1,pch=21)
```

```{r PGLS-Coleoptera-Hymenoptera, fig.cap="**Additional lineage-specific comparisons of egg volume vs ovariole number, phylogenetic residuals to body size, matching records at the genus level.** Weevils (Curculionoidea, left) were measured using elytra length and wasps (Hymenoptera, right) were measured using mesosoma width.", fig.width=6, fig.height=3, include=TRUE}
### Curculionoidea pgls
COL1_pgls <- COL1_body_egg_ON_genus %>% select(genus,logvol,logon,logbody,group) %>% rename(rank = genus, trait1 = logon, trait2 = logvol, indep = logbody)
theme_update(plot.title = element_text(hjust = 0.5))
col1 <- run_plotting_resid_pgls(COL1_pgls,tree,"COL1") + xlab("residual of ovariole number") +
  ylab("residual of egg volume") + ggtitle("Coleoptera: Curculionoidea")

### Apoidea pgls
HYM_pgls <- HYM_body_egg_ON_genus %>% select(genus,logvol,logon,logbody,group) %>% rename(rank = genus, trait1 = logon, trait2 = logvol, indep = logbody)
hym1 <- run_plotting_resid_pgls(HYM_pgls,tree,"HYM") + xlab("residual of ovariole number") +
  ylab(element_blank()) + ggtitle("Hymenoptera")

grid.arrange(col1,hym1,ncol=2)
```

```{r PGLS-body-plot, fig.cap="**Adult body size vs ovariole number.** Adult body weight compared at the genus level (left), and average adult body length compared at the family level (right).", fig.width=6, fig.height=3, include=TRUE}
### Gilbert body weight vs ON pgls
ON_body_gilbert_genus_pgls <- egg_ON_gilbert_genus %>% select(genus,logon,logweight,group) %>% mutate(rank = genus, trait1 = logon, trait2 = logweight)
gilbert_body <- run_plotting_pgls(ON_body_gilbert_genus_pgls,tree,"body_gilbert") + xlab(bquote('ovariole number,'~log[10])) +
  ylab(bquote('adult dry body weight (g),'~log[10]))

### PGLS body length vs ON, family averages
ON_body_family_pgls <- body_egg_ON_full %>% select(rank,logon,logbody,group) %>% rename(trait1 = logon, trait2 = logbody)
family_body <- run_plotting_pgls(ON_body_family_pgls,fam_tree,"body_egg_ON_family") + xlab(bquote('average ovariole number,'~log[10])) +
  ylab(bquote('average adult body length (mm),'~log[10]))

grid.arrange(gilbert_body,family_body,ncol=2) 
```

```{r PGLS-fecundity-genus}
### Gilbert fecundity vs ON, genus level
fecundity_genus_pgls <-  ON_gilbert_genus %>% select(genus,fecundity,on,group) %>% mutate(rank = genus, trait1 = on, trait2= fecundity) 
pdf("manuscript_figures/panels/FigFecundity.pdf",height=4,width=4,useDingbats = F)
  g1 <- ggplot(fecundity_genus_pgls,aes(x = trait1, y = trait2, color = group)) +
      geom_point() +
      scale_color_manual(values = mrk) +
      scale_x_log10() +
      scale_y_log10() +
      theme(legend.position = "none") +
      xlab('ovariole number') +
      ylab('lifetime fecundity')
  print(g1)
invisible(dev.off())
```

```{r PGLS-fecundity-name, fig.cap="**Lifetime fecundity vs ovariole number, matching records at the species level.**", fig.width=3, fig.height=3, include=TRUE}
### Gilbert fecundity vs ON, species level
fecundity_name_pgls <-  ON_gilbert_name %>% select(genus,logfecundity,logon,group) %>% mutate(rank = genus, trait1 = logon, trait2= logfecundity)
run_plotting_pgls(fecundity_name_pgls,tree,"fecundity_name") + xlab(bquote("ovariole number,"~log[10])) +
  ylab(bquote("lifetime fecundity,"~log[10]))
```

# Evolution of nurse cells

## Combining datasets

We used the descriptions of the mode of oogenesis recorded by BÃ¼ning[@buning1994insect]. This author catalogued the ovary morphology for 136 insect genera, categorizing them into four modes: those without nurse cells (panoistic), with nurse cells adjacent to each clonally related, developing oocyte (polytrophic meroistic), with all nurse cells located in the germarium (telotrophic meroistic), and a unique mode of oogenesis reported only in Strepsiptera (reduced polytrophic meroistic ovaries).

Of the 136 genera observed by BÃ¼ning, 70 are represented in the phylogeny used here[@church2019insect]. Another 36 come from families or orders that have representative genera in the phylogeny, and within which all observations have the same recorded mode of oogenesis, when more than one was recorded. Therefore, we used a substitute genus as the phylogenetic tip for these 36 groups, bringing the total overlap between dataset and phylogeny to 106 taxa.  

## Reconstructing evolutionary shifts in oogenesis mode

```{r reconstruct-ovary, include=TRUE, fig.cap = "**Ancestral state reconstruction of oogenesis mode, full phylogeny.** Scale bar indicates 100 million years. Gray = panoistic, red = telotrophic meroistic, cyan = polytrophic meroistic, black = unique meroistic mode found in Strepsiptera.", fig.width = 5, fig.height=9}
#source("analysis/reconstruct_ancestral_ovary_morphology_type.R")
load("analysis/mode_of_oogenesis_results_June2020/reconstruct_ancestral_ovary_morphology_type.RData")

# Extract estimated tip states
ovary_nums <- data.frame(num=as.numeric(as.factor(unique(ovary_table$ovary))),ovary=unique(ovary_table$ovary),stringsAsFactors=F)

est_tips <- data.frame("state1" = round(pp$tip.states,0)[,1],
            "state2" = round(pp$tip.states,0)[,2],
            "state3" = round(pp$tip.states,0)[,3],
            "state4" = round(pp$tip.states,0)[,4]) %>% 
      mutate(discrete = ifelse(state1 == 1,1,
                          ifelse(state2 == 1,2,
                            ifelse(state3 == 1,3,4))),
          genus = pp$phy$tip.label) %>% 
      select(genus,discrete) %>% 
      mutate("ovary_regime"=plyr::mapvalues(discrete,ovary_nums$num,ovary_nums$ovary))

ovary_cols <- c("black","#949494","#00c2c2","red")
names(ovary_cols) <- c("Mstar","P","PM","TM")

### Plot the full oogenesis mode ASR on the phylogeny
branch_cols <- c(as.integer(plyr::mapvalues(pp$phy$tip.label,est_tips$genus,est_tips$discrete)),pp$phy$node.label)
plot(pp$phy,show.tip.label=F,edge.width=1,edge.col=plyr::mapvalues(branch_cols[pp$phy$edge[,2]],seq(1:length(ovary_cols)),ovary_cols))
add.scale.bar()

### Plot just the reconstruction for the originally recorded 106 tips
pp_original_records <- drop.tip(pp$phy,tree_dataset %>% filter(!(genus %in% new_ovary_table$genus)) %>% pull(genus))
branch_cols <- c(as.integer(plyr::mapvalues(pp_original_records$tip.label,est_tips$genus,est_tips$discrete,warn_missing=F)),pp_original_records$node.label)
pdf("manuscript_figures/panels/FigBuning_Reconstruction.pdf",height=4.75,width=2.5,useDingbats = F)
  plot(pp_original_records,show.tip.label=F,edge.width=1,edge.col=plyr::mapvalues(branch_cols[pp_original_records$edge[,2]],seq(1:length(ovary_cols)),ovary_cols))
  add.scale.bar()
invisible(dev.off())
```

Using these 106 records for ovary type, we reconstructed the ancestral state at each node of the published phylogeny of `r length(pp$phy$tip.label)` insect genera using an equal-rates model that allows for missing data (Fig. \ref{fig:reconstruct-ovary}, R package corHMM,[@beaulieu2013identifying] version `r packageVersion('corHMM')`, function rayDISC, node.states = 'marginal').

## Oogenesis mode model comparison {#sec:oogenesis-model}

```{r ovary-model-comparison, include=TRUE}
# get corrected AIC values from results
get_aicc <- function(ou_res) {
  BM1 <- sapply(ou_res,function(x){x[[1]]$AICc})
  BMS <- sapply(ou_res,function(x){x[[2]]$AICc})
  OU1 <- sapply(ou_res,function(x){x[[3]]$AICc})
  OUM <- sapply(ou_res,function(x){x[[4]]$AICc})
  ouwie_res <- data.frame(BM1 = BM1, BMS = BMS, OU1 = OU1, OUM = OUM) %>%
        rowwise() %>%
        mutate(min = min(BM1,BMS,OU1,OUM)) %>%
        ungroup() %>%
        mutate(dBM1 = BM1 - min,
          dBMS = BMS - min,
          dOU1 = OU1 - min,
          dOUM = OUM - min,
          bm1_bms = (dBM1 - dBMS > 2),
          bm1_ou1 = (dBM1 - dOU1 > 2),
          bm1_oum = (dBM1 - dOUM > 2),
          ou1_oum = (dOU1 - dOUM > 2))
  return(ouwie_res)
}

# calculate the mean AICc value over the posterior distribution
get_mean_aicc <- function(res) {
  mean_ouwie_res <- res %>% select(BM1,BMS,OU1,OUM,dBM1,dBMS,dOU1,dOUM) %>%       
    group_by() %>%
    summarize_all("mean")
  return(mean_ouwie_res)
}

# count the number of AICc > 2
get_count_aicc <- function(res) {
  count_ouwie_res <- data.frame(bm1_bms = sum(res$bm1_bms),
          bm1_ou1 = sum(res$bm1_ou1),
          bm1_oum = sum(res$bm1_oum),
          ou1_oum = sum(res$ou1_oum))
  return(count_ouwie_res)
}

# use those functions to evaluate results for each trait of interest
load("analysis/mode_of_oogenesis_results_June2020/asym_ovary_model_comparison.RData") # egg asymmetry
asym_results <- ouwie_results
load("analysis/mode_of_oogenesis_results_June2020/curv_ovary_model_comparison.RData") # egg curvature
curv_results <- ouwie_results
load("analysis/mode_of_oogenesis_results_June2020/egg_ovary_model_comparison.RData") # egg volume
egg_results <- ouwie_results
load("analysis/mode_of_oogenesis_results_June2020/ar_ovary_model_comparison.RData") # egg aspect ratio
ar_results <- ouwie_results
load("analysis/mode_of_oogenesis_results_June2020/on_ovary_model_comparison.RData") # ovariole number
on_results <- ouwie_results

asym_res <- get_aicc(asym_results)
asym_mean_res <- get_mean_aicc(asym_res)
asym_count_res <- get_count_aicc(asym_res)

egg_res <- get_aicc(egg_results)
egg_mean_res <- get_mean_aicc(egg_res)
egg_count_res <- get_count_aicc(egg_res)

ar_res <- get_aicc(ar_results)
ar_mean_res <- get_mean_aicc(ar_res)
ar_count_res <- get_count_aicc(ar_res)

curv_res <- get_aicc(curv_results)
curv_mean_res <- get_mean_aicc(curv_res)
curv_count_res <- get_count_aicc(curv_res)

on_res <- get_aicc(on_results)
on_mean_res <- get_mean_aicc(on_res)
on_count_res <- get_count_aicc(on_res)

# count the number of taxa in each analysis
sample_size <- tibble("analysis name" = c("ovariole number","egg volume","egg aspect ratio","egg asymmetry","egg curvature"),
            "taxa" = c(nrow(on_results[[1]][[1]]$data),
              nrow(egg_results[[1]][[1]]$data),
              nrow(ar_results[[1]][[1]]$data),
              nrow(asym_results[[1]][[1]]$data),
              nrow(curv_results[[1]][[1]]$data)))

# write table
mean_table <- rbind(on_mean_res,egg_mean_res,ar_mean_res,asym_mean_res,curv_mean_res) %>%
  mutate("analysis name" = c("ovariole number","egg volume","egg aspect ratio","egg asymmetry","egg curvature")) %>%
  select("analysis name",BM1,BMS,OU1,OUM)
knitr::kable(mean_table, "latex",caption="Average corrected AIC (AICc) value from model comparison", booktabs = T, linesep = "") %>%
  kableExtra::kable_styling(latex_options = "striped") %>%
  kableExtra::column_spec(1, width = "10em") %>%
  kableExtra::kable_styling(latex_options = "hold_position")

```

```{r ovary-model-comparison-count, include=TRUE}
count_table <- rbind(on_count_res,egg_count_res,ar_count_res,asym_count_res,curv_count_res)  %>%
  mutate("analysis name" = c("ovariole number","egg volume","egg aspect ratio","egg asymmetry","egg curvature")) %>%
  select("analysis name",bm1_bms,bm1_ou1,bm1_oum,ou1_oum) %>%
  rename("BMS vs. BM1" = bm1_bms,
        "OU1 vs. BM1" = bm1_ou1,
        "OUM vs. BM1" = bm1_oum,
        "OUM vs. OU1" = ou1_oum) %>% 
  left_join(.,sample_size,by="analysis name")
knitr::kable(count_table, "latex",caption="Results of model comparison analysis over posterior distribution, showing the number of iterations out of 100 where the difference in model fit ($\\Delta$AICc) was greater than 2.", booktabs = T, linesep = "") %>%
  kableExtra::kable_styling(latex_options = "striped") %>%
  kableExtra::column_spec(1, width = "10em") %>%
  kableExtra::kable_styling(latex_options = "hold_position")
```

Using the ancestral state reconstruction of evolutionary shifts in the mode of oogenesis, we inferred the most likely mode of oogenesis for all nodes and unobserved extant tips in the phylogeny (Fig. \ref{fig:reconstruct-ovary}). We then compared the fit of models of trait evolution that take into account these shifts in oogenesis mode against those that do not. These comparisons were performed with the R package OUwie[@beaulieu2012modeling]  (version `r packageVersion('OUwie')`). 

Each analysis compared four models of evolution: single-rate Brownian Motion (BM1), multi-rate Brownian Motion (BMS), single-optimum Ornstein-Uhlenbeck (OU1), and an Ornstein-Uhlenbeck model with different optima for each mode of oogenesis (OUM). 

These comparisons were repeated 100 times over a posterior distribution of trees. At each iteration we selected a random representative trait record for each genus in the phylogeny, when multiple records were available. 


# Modeling rate of ovariole number change {#sec:arbutus}

## Parametric bootstrap of Brownian Motion model {#sec:bootstrap}

```{r arbutus, include=TRUE, fig.cap="**Bootstrap analysis of Brownian Motion model for ovariole number evolution, using the R package arbutus.** In each panel the red line represents the observed value and the black distribution represents the bootstrap simulation. See Section \\ref{sec:arbutus} for details on each parameter.",fig.width=9,fig.height=6}

# set up trait and tree for arbutus analysis
arbutus_tree <- drop.tip(tree,tree$tip.label[!(tree$tip.label %in% ON$genus)])
arbutus_trait <- ON %>% filter(genus %in% arbutus_tree$tip.label) %>%
  group_by(genus) %>% slice(1L)
arbutus_trait_vec <- arbutus_trait %>% pull(logon)
names(arbutus_trait_vec) <- arbutus_trait$genus

# fit a Brownian Motion model
fit_bm <- arbutus_trait_vec[order(match(names(arbutus_trait_vec),arbutus_tree$tip.label))] %>%
  fitContinuous(dat = ., phy =arbutus_tree, model="BM")

# run a parametric bootstrap of the model with arbutus
arbutus_fit_bm <- arbutus(fit_bm)

# plot the results
arbutus_names <- c("Mean of squared contrasts","Variation of contrast absolute values","Slope of contrast absolute values against variances","Slope of contrast absolute value against trait value","Slope of contrast absolute value against node age","D-statistic, difference from normality")
names(arbutus_fit_bm$obs) <- arbutus_names
names(arbutus_fit_bm$sim) <- arbutus_names
plot(arbutus_fit_bm,legend=F)
```

We evaluated the fit of a Brownian Motion (BM) model for ovariole number evolution using the R package arbutus[@pennell2015model], version `r packageVersion('arbutus')`  (Fig. \ref{fig:arbutus}). In this approach, a BM model is fit to the data (R package geiger[@harmon2007geiger], version `r packageVersion('geiger')`), and the resulting parameters of the model are used to simulate 1000 new datasets. Six statistical parameters are used to compare the phylogenetic contrasts of the observed data to the simulated data, and their interpretations are as follows[@pennell2015model]:


1. _Mean of squared contrasts_. The rate of evolution of ovariole number can be well estimated by the Brownian Motion model (the observed value falls within the null distribution).
2. _Coefficient of variation of the absolute value of the contrasts._ There is substantially more variation in contrasts than expected by chance, indicating heterogeneity in the rate of evolution beyond what a single-rate Brownian Motion model predicts (the observed value falls well outside the null distribution).
3. _Slope of a linear model fitted to the absolute value of the contrasts against their expected variances._ Contrasts are larger than expected on short branches in the phylogenetic tree, resulting in a negative slope. This could be explained by error in estimation of branch lengths.
4. _Slope of a linear model fitted to the absolute value of the contrasts against the ancestral state at the corresponding node._ The number of ovarioles is more correlated with contrast values than would be expected by chance. Phylogenetic nodes with a low ovariole number experience lower rates of evolution.
5. _Slope of a linear model fitted to the absolute value of the contrasts against node depth._ Contrast values are not correlated with time, falling within the null distribution. Therefore the rate of ovariole number change is not increasing or decreasing over time.
6. _The D statistic from a Kolmogorov-Smirnov test comparing the distribution of contrasts to an expected normal distribution._ The data do not fit a normal distribution of contrasts well, suggesting there are likely non-Brownian motion based processes at play (e.g. jump-diffusion processes).

## Assessing rate heterogeneity {#sec:BAMM}

```{r BAMM-analysis-setup}
### Select a continuous trait to analyze
trait_of_interest <- "on"
dat_of_interest <- ON

### name the analysis
name <- paste("BAMM",trait_of_interest,sep="_")

### Set up the data frame for analysis
dat <- dat_of_interest[sample(nrow(dat_of_interest)),] %>% # shuffle the dataset
  mutate(raw_trait = eval(parse(text = trait_of_interest))) %>%  # pull out genus and trait
  filter(!(is.na(raw_trait))) %>% select(genus,raw_trait) %>% # remove missing data
  group_by(genus) %>% summarize(mean_trait = mean(raw_trait)) %>% # calculate mean per genus
  mutate(trait = log10(mean_trait)) # transform the data

### Format the data frame for BAMM
bmm_red <- dat %>% mutate(taxa = genus) %>%
    filter(genus %in% tree$tip.label) %>%
    select(genus,trait)

### Use only taxa with phenotypic data
bmm <- drop.tip(tree,setdiff(tree$tip.label,bmm_red$genus))

### Write the files for BAMM
#write.table(bmm_red,file="bamm_trait.txt",sep="\t",row.names=F,col.names=F,quote=F)
#write.tree(bmm,file="bamm_tree.tre")

# Select appropriate priors
#priors <- setBAMMpriors(phy = bmm, traits = "bamm_trait.txt", outfile=NULL)
#priors[1] <- 10

#params <- c(treefile = 'bamm_tree.tre',
#        traitfile = 'bamm_trait.txt',
#        numberOfGenerations = '2000000000', # use 10^9 generations
#        mcmcWriteFreq = '50000',
#        eventDataWriteFreq = '1000000', # sample every 10^6
#        outName = 'bamm',
#        overwrite = '1',
#        printFreq = '1000000',
#        priors)

#generateControlFile('bamm.config', type = 'trait', params = as.list(params))

#### OUTSIDE OF R, RUN BAMM
#### bamm -c bamm.config
```

Given the result that our dataset contains substantial rate heterogeneity, we identified regions of the tree with high and low rates of ovariole number evolution using the software BAMM[@rabosky2014automatic], version 2.5.0. For this analysis, we calculated the average ovariole number for each genus in the insect phylogeny[@church2019insect]. Average ovariole number was log~10~ transformed, and the tree was filtered to include only tips for which there were corresponding ovariole number data (sample size = `r length(bmm$tip.label)`). We used the R package BAMMtools (version `r packageVersion('BAMMtools')`)[@rabosky2014bamm] to select priors, and ran BAMM for the maximum number of generations ($2*10^{9}$), sampling every 10^6^ generations.

```{r BAMM-analysis-convergence, include=TRUE, fig.cap="Convergence of trait diversification rate analysis", fig.width=10, fig.height=5}
### Assess convergence
mcmcout <- read.csv("analysis/BAMM_results_June2020/bamm_mcmc_out.txt", header=T) 

#pdf(file=paste(name,"_convergence.pdf",sep=""),width=10,height=5)
  plot(mcmcout$logLik ~ mcmcout$generation,pch=16,cex=0.5,xlab="MCMC generation",ylab="Log-likelihood")
#dev.off()
```

```{r BAMM-analysis-burnin, include=TRUE, fig.cap="**Comparison of burn-in proportions.** The burn-in proportion that maximized the effective size was used in subsequent analyses.", fig.width=10, fig.height=5}
# calculate the effective size of the logL using different burn in values
bi <- seq(0.01,0.99,by=0.001) 
es_logl <- sapply(bi,function(x){effectiveSize(mcmcout[(x * nrow(mcmcout)):nrow(mcmcout),]$logLik)})

burnin <- bi[which.max(es_logl)]

#pdf(file=paste(name,"_effectiveSize.pdf",sep=""),width=10,height=5)
  plot(y=es_logl,x=bi,pch=16,cex=0.5,xlab="Proportion designated as burn-in",ylab="Effective sample size of log-likelihood")
#dev.off()

burnstart <- floor(burnin * nrow(mcmcout))
postburn <- mcmcout[burnstart:nrow(mcmcout), ]

es_nshifts <- effectiveSize(postburn$N_shifts) # this should be >200
es_logl <- effectiveSize(postburn$logLik) # this should be >200, though it has never gotten close
#es_nshifts
#es_logl
```

Convergence was evaluated both visually (Fig. \ref{fig:BAMM-analysis-convergence}) and numerically by comparing the effective sample size for number of shifts and log-likelihood to the standard recommended by the software (>200). We determined the most appropriate burn-in proportion to use by finding the maximum effective sample size of the log-likelihood across an array of possible burn-in proportions (Fig. \ref{fig:BAMM-analysis-burnin}). Running BAMM for the maximum possible number of generations and selecting the optimum burn-in\ref{fig:BAMM-analysis-burnin} resulted in an effective size for the number of shifts of `r round(es_nshifts,2)`, and for log-likelihood of `r round(es_logl,2)`. Repeated BAMM analyses showed similar distributions of high and low rate regimes, indicating the implications for ovariole number evolution are robust to uncertainty in rate estimates. See Supplemental Methods Section \ref{sec:BAMM} for details.

```{r BAMM-anlaysis-event-data}
### Read in the BAMM results and plot the shifts on the tree
bamm_tree <- ladderize(read.tree(file="analysis/BAMM_results_June2020/bamm_tree.tre")) 

# get event data
edata <- getEventData(bamm_tree, eventdata = paste("analysis/BAMM_results_June2020/bamm_event_data.txt",sep=""), burnin=burnin, type="trait")
```

```{r BAMM-analysis-results, include=TRUE, fig.cap="**Best rate shift configuration from BAMM trait diversification analysis on ovariole number.** Purple = low rate of evolution, yellow = high rate of evolution.", fig.width=5, fig.height=10}
# plot the best tree as a phylorate plot
#pdf(file=paste(name,"_ratesplot.pdf",sep=""),width=10,height=12)
  pb_legend <- png::readPNG("analysis/datafiles/BAMM_legend.png")
  grid.arrange(ggplotify::as.grob(~plot(edata,lwd=2,legend=F,labels=F,cex=0.2,logcolor=T,breaksmethod="jenks",pal=c("#0D0887", "#CC4678", "#F0F921"))),rasterGrob(pb_legend),ncol=2,widths=c(10,1)) 
#dev.off()
```

```{r BAMM-anlaysis-mean-rates, include=TRUE, fig.cap="**Distribution of trait diversification rates.** Dotted line shows threshold used to assigned rate regimes.", fig.width=10, fig.height=5}
### Identify fixed lineages
# build a tree with branch lengths encoded as mean rate
brlen <- getMeanBranchLengthTree(edata)

# choose a low rate threshold 
lowrate <- 1e-04

# plot rates
#pdf(file=paste(name,"_meanrates.pdf",sep=""),width=10,height=5)
  hist(log10(brlen$phy$edge.length),breaks=200,col="black",xlab=bquote('Mean branch rate of trait diversification,'~log[10]),main="")
  abline(v = log10(lowrate), col="black", lwd=3, lty=2)
#dev.off()
```

The best configuration of rate shift regimes shows multiple independent clades with very low rates of evolution (Fig. \ref{fig:BAMM-analysis-results}). Visualizing the distribution of mean rates along branches revealed a discontinuous distribution, with one peak at a moderate rate of evolution and another at an extremely low rate, separated from the first peak by over six orders of magnitude (Fig. \ref{fig:BAMM-anlaysis-mean-rates}). We used this visualization to establish a threshold (`r lowrate`) for assigning a binary rate regime to each node in the phylogeny, categorizing them as above (variable) or below (invariant) a threshold that separates these two peaks.

```{r BAMM-analysis-mainfig}
# bin the rates into above or below the threshold
regimes <- data.frame(node=brlen$phy$edge[,2], # use the terminal nodes of each edge as the identifier
          brlen=brlen$phy$edge.length) %>% # pull the branch length = mean rate
      mutate(regime = ifelse(brlen<lowrate,"below","above")) # bin rates
mrca_node <- data.frame(node=getMRCA(bamm_tree,bamm_tree$tip.label),brlen=NA,regime="above") # set the root node as above
regimes <- rbind(regimes,mrca_node) %>% arrange(node) # add root

### Plot the tree with binned fixed lineages
# start with the bamm trees 
node_tree <- bamm_tree

# label the internal nodes to match the regime data
node_tree$node.label <- regimes %>% filter(node > length(node_tree$tip.label)) %>% pull(regime)

# build vector of branch colors, mapped to the order of terminal nodes in the tree
cols <- c("dark gray","black")
branch_cols <- plyr::mapvalues(regimes$regime[node_tree$edge[,2]],c("above","below"),cols)

### Assess trait evolution using the rate regimes
on_reg <- regimes[1:length(node_tree$tip.label),] %>% mutate(genus = node_tree$tip.label) %>% # pull out tips
  left_join(.,(ON %>% select(genus,on,group)),by="genus") %>% # add in trait data
  rename(rate = "brlen") %>% # recode for geiger
  select(genus,on,regime,rate,group)

# plot panels for figure:Diversity
size <- 0.8

# phylogeny
p0 <- ggtree(node_tree,size=0.35,color="dark gray") + scale_color_manual(values=mrk) + scale_fill_manual(values=mrk) + scale_y_reverse()  + theme_tree2()
p02 <- facet_plot(p0,panel="ovariole number",data=as.data.frame(on_reg),geom=geom_point,aes(x = log10(on),color = group),size=size,pch=16) + theme(legend.position = "none")

# add regimes
p <- ggtree(node_tree,size=0.35) %<+% (regimes) + aes(color=regime) + scale_color_manual(values=mrk) + scale_fill_manual(values=mrk) + scale_y_reverse()  + theme_tree2()
p2 <- facet_plot(p,panel="ovariole number",data=as.data.frame(on_reg),geom=geom_point,aes(x = log10(on),color = group),size=size,pch=16) + theme(legend.position = "none")

# change the widths of the phylogeny plot
p02t <- ggplot_gtable(ggplot_build(p02))
p02t$widths[7] = 2*p02t$widths[7] # in this case it was colmun 7 - reduce the width by a half

p2t <- ggplot_gtable(ggplot_build(p2))
p2t$widths[7] = 2*p2t$widths[7] # in this case it was colmun 7 - reduce the width by a half

# jitter plot
g1 <- ggplot(ON,aes(y = factor(group,levels=rev(group_levels)), x= on, color = group)) + geom_jitter(size=size,alpha=0.8,pch=16) + scale_color_manual(values=mrk) + scale_x_log10() + theme(legend.position = "none",axis.text.y=element_blank()) + xlab('ovariole number')
# histogram
g2 <- ggplot(ON,aes(x = on, fill=group)) + geom_histogram(bins=125) + scale_fill_manual(values=mrk) + scale_x_log10() + theme(legend.position = "none") + xlab('ovariole number')
g3 <- ggplot(ON %>% filter(genus %in% node_tree$tip.label),aes(x = on, fill=group)) + geom_histogram(bins=125) + scale_fill_manual(values=mrk) + scale_x_log10() + theme(legend.position = "none") + xlab('ovariole number')

pdf("manuscript_figures/panels/FigDiversity_Phylogeny.pdf",height=7,width=3.5,useDingbats = F)
  print(p2)
invisible(dev.off())
pdf("manuscript_figures/panels/FigDiversity_Phylogeny_Wide.pdf",height=7,width=3.5,useDingbats = F)
  grid.draw(p2t)
invisible(dev.off())
pdf("manuscript_figures/panels/FigDiversity_Phylogeny_NoRegime.pdf",height=7,width=3.5,useDingbats = F)
  grid.draw(p02t) 
invisible(dev.off())
pdf("manuscript_figures/panels/FigDiversity_Jitter.pdf",height=3.25,width=3.25,useDingbats = F)
  print(g1)
invisible(dev.off())
pdf("manuscript_figures/panels/FigDiversity_Histogram.pdf",height=3.25,width=4,useDingbats = F)
  print(g2)
invisible(dev.off())
pdf("manuscript_figures/panels/FigDiversity_Histogram_Short.pdf",height=1.5,width=2.75,useDingbats = F)
  print(g3)
invisible(dev.off())
```

## Rate model comparison

```{r BM-model-comparison}
# set up the data frame to compare models
ouwie_on_reg <- on_reg[sample(nrow(on_reg)),] %>% # shuffle the dataset
    group_by(genus) %>% slice(1L) %>% ungroup() %>% # choose one representative per genus
    mutate(species = genus,discrete = regime,continuous = log10(on)) %>% #transform the data
    select(species,discrete,continuous) %>% 
    as.data.frame()

### Compare BM models of evolution with and without regimes
BMS <- OUwie(node_tree,ouwie_on_reg,model="BMS",simmap.tree=FALSE,diagn=FALSE,root.station=F) #rates per regime
BM1 <- OUwie(node_tree,ouwie_on_reg,model="BM1",simmap.tree=FALSE,diagn=FALSE,root.station=F) #one rate
```

We tested whether a BM model of evolution that incorporates the binary state (variable or invariant) as independent rate regimes can better explain the distribution of ovariole numbers than a single rate BM model, by comparing model fit using the R package OUwie (version `r packageVersion('OUwie')`)[@beaulieu2012modeling]. We find that a multi-rate model is significantly favored over a single-rate model ($\Delta$ AICc `r round(BM1$AICc - BMS$AICc,2)`).

## Comparing rates of trait diversification

```{r BAMM-egg-setup}
tree <- genus_mcc_tree
trait_of_interest <- "vol"
dat_of_interest <- egg_database
dat <- dat_of_interest[sample(nrow(dat_of_interest)),] %>% # shuffle the dataset
  mutate(raw_trait = eval(parse(text = trait_of_interest))) %>%  # pull out genus and trait
  filter(!(is.na(raw_trait))) %>% select(genus,raw_trait) %>% # remove missing data
  group_by(genus) %>% summarize(mean_trait = mean(raw_trait)) %>% # calculate mean per genus
  mutate(trait = log10(mean_trait)) # transform the data

bmm_red <- dat %>% mutate(taxa = genus) %>%
    filter(genus %in% tree$tip.label) %>%
    select(genus,trait)
bmm <- drop.tip(tree,setdiff(tree$tip.label,bmm_red$genus))

#write.table(bmm_red,file="egg_bamm_trait.txt",sep="\t",row.names=F,col.names=F,quote=F)
#write.tree(bmm,file="egg_bamm_tree.tre")

#priors <- setBAMMpriors(phy = bmm, traits = "egg_bamm_trait.txt", outfile=NULL)
#priors[1] <- 10
#params <- c(treefile = 'egg_bamm_tree.tre',
#        traitfile = 'egg_bamm_trait.txt',
#        numberOfGenerations = '2000000000', # use 10^9 generations
#        mcmcWriteFreq = '50000',
#        eventDataWriteFreq = '1000000', # sample every 10^6
#        outName = 'egg_bamm',
#        overwrite = '1',
#        printFreq = '1000000',
#        priors)

#generateControlFile('egg_bamm.config', type = 'trait', params = as.list(params))

#### OUTSIDE OF R, RUN BAMM
#### bamm -c egg_bamm.config
```

```{r BAMM-egg-read}
egg_mcmcout <- read.csv("analysis/BAMM_results_June2020/egg_bamm_mcmc_out.txt", header=T)

egg_bi <- seq(0.01,0.99,by=0.001)
egg_es_logl <- sapply(egg_bi,function(x){effectiveSize(egg_mcmcout[(x * nrow(egg_mcmcout)):nrow(egg_mcmcout),]$logLik)})

egg_burnin <- egg_bi[which.max(egg_es_logl)]
egg_burnstart <- floor(egg_burnin * nrow(egg_mcmcout))
egg_postburn <- egg_mcmcout[egg_burnstart:nrow(egg_mcmcout), ]
egg_es_nshifts <- effectiveSize(egg_postburn$N_shifts)
egg_es_logl <- effectiveSize(egg_postburn$logLik)

egg_bamm_tree <- ladderize(read.tree(file="analysis/BAMM_results_June2020/egg_bamm_tree.tre"))
egg_edata <- getEventData(egg_bamm_tree, eventdata = paste("analysis/BAMM_results_June2020/egg_bamm_event_data.txt",sep=""), burnin=egg_burnin, type="trait")
```

```{r BAMM-egg-plot, include=TRUE, fig.cap="**Best rate shift configuration from BAMM trait diversification analysis on egg volume.** Purple = low rate of evolution, yellow = high rate of evolution.", fig.width=5, fig.height=10}
  pb_legend <- png::readPNG("analysis/datafiles/BAMM_legend.png")
  grid.arrange(ggplotify::as.grob(~plot(egg_edata,lwd=2,legend=F,labels=F,cex=0.2,logcolor=T,breaksmethod="jenks",pal=c("#0D0887", "#CC4678", "#F0F921"))),rasterGrob(pb_legend),ncol=2,widths=c(10,1))
```

```{r combine-rates, include=TRUE, fig.cap="**Rates of trait diversification of egg volume and ovariole number.** Points are colored by phylogenetic groups shown in Fig 1a. Dotted line shows threshold used to assigned rate regimes.", fig.width = 7, fig.height = 3.5}
egg_brlen <- getMeanBranchLengthTree(egg_edata)
egg_regimes <- data.frame(node=egg_brlen$phy$edge[,2],brlen=egg_brlen$phy$edge.length)
 
egg_reg <- egg_regimes[1:length(egg_bamm_tree$tip.label),] %>% mutate(genus = egg_bamm_tree$tip.label) %>% # pull out tips
  left_join(.,(egg_database %>% select(genus,vol,group)),by="genus") %>% # add in trait data
  rename(egg_rate = "brlen") %>% # recode for geiger
  select(genus,vol,egg_rate) %>% na.omit()

on_egg_reg <- on_reg[sample(nrow(on_reg)),] %>%
          group_by(genus) %>%
          slice(1L) %>%
          left_join(.,egg_reg[sample(nrow(egg_reg)),] %>%
              group_by(genus) %>%
              slice(1L),by="genus")

ggplot(on_egg_reg,aes(x = rate, y = egg_rate, color= group)) + geom_point() +
  scale_color_manual(values = mrk) +
  scale_x_log10(label=scientific_10) +
  scale_y_log10(label=scientific_10) +
  theme(legend.position = "none") +
  xlab("rate of ovariole number diversification") +
  ylab("rate of egg volume diversification") +
  geom_vline(xintercept=1e-04,linetype="dashed")
```

We assessed the rate of egg volume diversification across insects using the same method of assessing rate heterogeneity as described in Section \ref{sec:BAMM} (Fig. \ref{fig:BAMM-egg-plot}). This analysis converged in $20*10^8$ generations (the effective size for the number of shifts and log-likelihood were >200, `r round(egg_es_nshifts,2)` and `r round(egg_es_logl,2)` respectively). We compared the correlation between the rates of trait diversification for ovariole number and egg volume by matching the mean rate predicted for each insect genus (the tips of the phylogeny from the BAMM analyses, Fig. \ref{fig:combine-rates}).


\pagebreak
# References

```{r save, cache=FALSE}
save.image("manuscript_supplement.RData")
```

